crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,781 @@
+        if !qargs1.iter().any(|e| qargs2.contains(e)) && !cargs1.iter().any(|e| cargs2.contains(e))
+        {
+            return Some(true);
+        }
+
+        if qargs1.len() > max_num_qubits as usize || qargs2.len() > max_num_qubits as usize {
+            return Some(false);
+        }
+
+        if SUPPORTED_OP.contains(&op1.op().name()) && SUPPORTED_OP.contains(&op2.op().name()) {
+            return None;
+        }
+
+        if self.is_commutation_skipped(op1) || self.is_commutation_skipped(op2) {
+            return Some(false);
+        }
+
+        None
+    }
+
+    fn get_op(py: Python, gate: &PyObject) -> Option<Array2<Complex64>> {
+        Some(
+            QI_OPERATOR
+                .get_bound(py)
+                .call1((gate,))
+                .ok()?
+                .getattr(intern!(py, "data"))
+                .ok()?
+                .extract::<PyReadonlyArray2<Complex64>>()
+                .ok()?
+                .as_array()
+                .to_owned(),
+        )
+    }
+
+    fn is_commutation_skipped(&self, instr: &CircuitInstruction) -> bool {
+        let op = instr.op();
+        op.directive() || SKIPPED_NAMES.contains(&op.name()) || instr.is_parameterized()
+    }
+
+    fn get_relative_placement(
+        first_qargs: &[usize],
+        second_qargs: &[usize],
+    ) -> SmallVec<[Option<Qubit>; 2]> {
+        let qubits_g2: HashMap<_, _> = second_qargs
+            .iter()
+            .enumerate()
+            .map(|(i_g1, q_g1)| (q_g1, Qubit(i_g1 as u32)))
+            .collect();
+
+        first_qargs
+            .iter()
+            .map(|q_g0| qubits_g2.get(q_g0).copied())
+            .collect()
+    }
+
+    fn clear_cache(&mut self) {
+        self.cache.clear();
+        self.current_cache_entries = 0;
+        self._cache_miss = 0;
+        self._cache_hit = 0;
+        self._cache_hit = 0;
+    }
+}
+
+#[derive(Clone)]
+#[pyclass]
+pub struct CommutationLibrary {
+    pub library: HashMap<(String, String), CommutationLibraryEntry>,
+}
+
+impl CommutationLibrary {
+    fn check_commutation_entries(
+        &self,
+        first_op: &OperationRef,
+        first_qargs: &[usize],
+        second_op: &OperationRef,
+        second_qargs: &[usize],
+    ) -> Option<bool> {
+        match self
+            .library
+            .get(&(first_op.name().to_string(), second_op.name().to_string()))
+        {
+            Some(CommutationLibraryEntry::Commutes(b)) => Some(*b),
+            Some(CommutationLibraryEntry::QubitMapping(qm)) => qm
+                .get(&CommutationChecker::get_relative_placement(
+                    first_qargs,
+                    second_qargs,
+                ))
+                .copied(),
+            _ => None,
+        }
+    }
+}
+
+impl ToPyObject for CommutationLibrary {
+    fn to_object(&self, py: Python) -> PyObject {
+        self.library.to_object(py)
+    }
+}
```

Comment:
I wonder what the use case for this method is? I understand that `CommutationLibrary` is already marked as `#[pyclass]` therefore this would convert to a python instance of `CommutationLibrary`.

This could potentially create conflicts when this is sent back to python as you could end up with either an instance of `dict` or `CommutationLibrary`. You probably do not want something like this.

Reply:
I think I got a compiler error when trying to serialize without this code...

Reply:
Yup 
```
out_dict.set_item("library", self.library.clone())?;
the trait `pyo3::ToPyObject` is not implemented for `CommutationLibrary
```

Reply:
Oh I see. For this I believe you could do `self.library.clone().into_py(py)`. And it works similarly.

Reply:
This does not appear to work out, it incurs the error `error[E0599]: the method `into_py` exists for struct `HashMap<(String, String), CommutationLibraryEntry>`, but its trait bounds were not satisfied
`

Reply:
I think I see the issue here, the thing is that for `CommutationLibraryEntry` only the `ToPyObject` trait was implemented and not the `IntoPy` one. `IntoPy` would consume whatever is being sent over, while `ToPyObject` generates a clone. Which is why it doesn't work with what I sent. 

You might want to consider making an implementation of `IntoPy<PyObject>` and then make the implementation to `ToPyObject` be just `clone.into_py(py)`.